### PR TITLE
[Merged by Bors] - feat: have cache use temporary files and renaming to avoid broken files

### PR DESCRIPTION
**Pull request by kim-em**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Cache/Requests.lean [hunk 1]
```diff
@@ -44,20 +44,26 @@ def mkGetConfigContent (hashMap : IO.HashMap) : IO String := do
     -- A backslash preceding any other letter is ignored.
     -- ```
     -- If this becomes an issue we can implement the curl spec.
+
+    -- Note we append a '_' to the filenames here,
+    -- which `downloadFiles` then removes when the download is successful.
     pure $ acc ++ s!"url = {← mkFileURL fileName none}\n-o {
-      (IO.CACHEDIR / fileName).toString.quote}\n"
+      (IO.CACHEDIR / (fileName ++ "_")).toString.quote}\n"
```
Comment (jcommelin) [inline Cache/Requests.lean:51]: How about `.part` as suffix. I think other programs also use that, when downloading. It's a bit more descriptive.

Comment (semorrison) [inline Cache/Requests.lean:51]:
```suggestion
      (IO.CACHEDIR / (fileName ++ ".part")).toString.quote}\n"
```

File: Cache/Requests.lean [hunk 2]
```diff
@@ -44,20 +44,26 @@ def mkGetConfigContent (hashMap : IO.HashMap) : IO String := do
...
 
 /-- Calls `curl` to download a single file from the server to `CACHEDIR` (`.cache`) -/
 def downloadFile (hash : UInt64) : IO Bool := do
   let fileName := hash.asLTar
   let url ← mkFileURL fileName none
   let path := IO.CACHEDIR / fileName
+  let fileName_ := fileName ++ "_"
+  let path_ := IO.CACHEDIR / fileName_
```
Comment (semorrison) [inline Cache/Requests.lean:59]:
```suggestion
  let partPath := IO.CACHEDIR / fileName_
```

File: Cache/Requests.lean [hunk 3]
```diff
@@ -44,20 +44,26 @@ def mkGetConfigContent (hashMap : IO.HashMap) : IO String := do
...
   let out ← IO.Process.output
-    { cmd := (← IO.getCurl), args := #[url, "--fail", "--silent", "-o", path.toString] }
+    { cmd := (← IO.getCurl), args := #[url, "--fail", "--silent", "-o", path_.toString] }
```
Comment (semorrison) [inline Cache/Requests.lean:61]:
```suggestion
    { cmd := (← IO.getCurl), args := #[url, "--fail", "--silent", "-o", partPath.toString] }
```

File: Cache/Requests.lean [hunk 4]
```diff
@@ -44,20 +44,26 @@ def mkGetConfigContent (hashMap : IO.HashMap) : IO String := do
...
   if out.exitCode = 0 then
+    IO.FS.rename path_ path
     pure true
   else
-    IO.FS.removeFile path
+    IO.FS.removeFile path_
```
Comment (semorrison) [inline Cache/Requests.lean:66]:
```suggestion
    IO.FS.removeFile partPath
```

File: Cache/Requests.lean [hunk 5]
```diff
@@ -44,20 +44,26 @@ def mkGetConfigContent (hashMap : IO.HashMap) : IO String := do
...
   let fileName := hash.asLTar
   let url ← mkFileURL fileName none
   let path := IO.CACHEDIR / fileName
+  let fileName_ := fileName ++ "_"
+  let path_ := IO.CACHEDIR / fileName_
   let out ← IO.Process.output
-    { cmd := (← IO.getCurl), args := #[url, "--fail", "--silent", "-o", path.toString] }
+    { cmd := (← IO.getCurl), args := #[url, "--fail", "--silent", "-o", path_.toString] }
   if out.exitCode = 0 then
+    IO.FS.rename path_ path
```
Comment (semorrison) [inline Cache/Requests.lean:63]:
```suggestion
    IO.FS.rename partPath path
```

File: Cache/Requests.lean [hunk 6]
```diff
@@ -44,20 +44,26 @@ def mkGetConfigContent (hashMap : IO.HashMap) : IO String := do
...
     pure $ acc ++ s!"url = {← mkFileURL fileName none}\n-o {
-      (IO.CACHEDIR / fileName).toString.quote}\n"
+      (IO.CACHEDIR / (fileName ++ "_")).toString.quote}\n"
 
 /-- Calls `curl` to download a single file from the server to `CACHEDIR` (`.cache`) -/
 def downloadFile (hash : UInt64) : IO Bool := do
   let fileName := hash.asLTar
   let url ← mkFileURL fileName none
   let path := IO.CACHEDIR / fileName
+  let fileName_ := fileName ++ "_"
```
Comment (semorrison) [inline Cache/Requests.lean:58]:
```suggestion
  let partFileName := fileName ++ ".part"
```

File: Cache/Requests.lean [hunk 7]
```diff
@@ -79,7 +85,11 @@ def downloadFiles (hashMap : IO.HashMap) (forceDownload : Bool) (parallel : Bool
         if !line.isEmpty then
           let result ← IO.ofExcept <| Lean.Json.parse line
           match result.getObjValAs? Nat "http_code" with
-          | .ok 200 => success := success + 1
+          | .ok 200 =>
+            if let .ok fn := result.getObjValAs? String "filename_effective" then
+              if (← System.FilePath.pathExists fn) && fn.endsWith "_" then
+                IO.FS.rename fn (fn.dropRight 1)
```
Comment (semorrison) [inline Cache/Requests.lean:91]:
```suggestion
              if (← System.FilePath.pathExists fn) && fn.endsWith ".part" then
                IO.FS.rename fn (fn.dropRight 5)
```

File: Cache/Requests.lean [hunk 8]
```diff
@@ -44,20 +44,26 @@ def mkGetConfigContent (hashMap : IO.HashMap) : IO String := do
     -- A backslash preceding any other letter is ignored.
     -- ```
     -- If this becomes an issue we can implement the curl spec.
+
+    -- Note we append a '_' to the filenames here,
```
Comment (sgouezel) [inline Cache/Requests.lean:48]:
```suggestion
    -- Note we append a '.part' to the filenames here,
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**